protocol/auditlog.go

```diff
@@ -0,0 +1,205 @@
+	}
+	return initSTRHash
+}
+
+func newDirectoryHistory(dirName string, signKey sign.PublicKey) *directoryHistory {
+	h := new(directoryHistory)
+	h.dirName = dirName
+	h.signKey = signKey
+	h.snapshots = make(map[uint64]*DirSTR)
+	h.latestSTR = nil
+	return h
+}
+
+// NewAuditLog constructs a new ConiksAuditLog. It creates an empty
+// log; the auditor will add an entry for each CONIKS directory
+// the first time it observes an STR for that directory.
+func NewAuditLog() ConiksAuditLog {
+	l := make(map[[crypto.HashSizeByte]byte]*directoryHistory)
+	return l
+}
+
+// IsKnownDirectory checks to see if an entry for the directory
+// (indexed by the hash of its initial STR dirInitHash) exists
+// in the audit log l. IsKnownDirectory() does not
+// validate the entries themselves. It returns true if an entry exists,
+// and false otherwise.
+func (l ConiksAuditLog) IsKnownDirectory(dirInitHash [crypto.HashSizeByte]byte) bool {
+	h := l[dirInitHash]
+	if h != nil {
+		return true
+	}
+	return false
+}
+
+// updateLatestSTR inserts a new STR into a directory history;
+// assumes the STR has been validated by the caller
+func (h *directoryHistory) updateLatestSTR(newLatest *DirSTR) {
+	h.snapshots[newLatest.Epoch] = newLatest
+	h.latestSTR = newLatest
+}
+
+// Insert creates a new directory history for the key directory addr
+// and inserts it into the audit log l.
+// The directory history is initialized with the key directory's
+// signing key signKey, and a list of STRs representing the
+// directory's STR history so far (as a map of epochs to STRs).
+// Insert() returns an ErrAuditLog if the auditor attempts to create
+// a new history for a known directory, an ErrMalformedDirectoryMessage
+// if oldSTRs is malformed, and nil otherwise.
+// Insert() only creates the initial entry in the log for addr. Use Update()
+// to insert newly observed STRs for addr in subsequent epochs.
+// FIXME: pass Response message as param
+// masomel: will probably want to write a more generic function
+// for "catching up" on a history in case an auditor misses epochs
+func (l ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	hist map[uint64]*DirSTR) error {
+
+	// make sure we're getting an initial STR at the very least
+	if len(hist) < 1 && hist[0].Epoch != 0 {
+		return ErrMalformedDirectoryMessage
+	}
+
+	// compute the hash of the initial STR
+	dirInitHash := computeInitSTRHash(hist[0])
+
+	// error if we want to create a new entry for a directory
+	// we already know
+	if l.IsKnownDirectory(dirInitHash) {
+		return ErrAuditLog
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(addr, signKey)
+
+	startEp := uint64(0)
+	endEp := uint64(len(hist))
+
+	// add each STR into the history
+	// This loop automatically catches if hist is malformed
+	// (i.e. hist is missing and epoch)
+	// FIXME: verify the consistency of each new STR before inserting
+	// into the audit log
+	for ep := startEp; ep < endEp; ep++ {
```

Comment:
I think that the initial STR should be in `snapshots` since it's also pinned in the client, and the client verifies its directory's history with this pinned STR. So the auditor should mirror the entire history for each directory that it tracks. But I do like your idea that `newDirectoryHistory` should take the initial STR as a parameter.